src/stream/src/executor/exchange/input.rs

```diff
@@ -64,23 +66,80 @@ pub struct LocalInput {
 }
 type LocalInputStreamInner = impl MessageStream;
 
+async fn process_msg<'a>(
+    msg: DispatcherMessage,
+    get_mutation_subscriber: impl for<'b> FnOnce(
+            &'b DispatcherBarrier,
+        )
+            -> &'a mut mpsc::UnboundedReceiver<crate::task::SubscribeMutationItem>
+        + 'a,
+) -> StreamExecutorResult<Message> {
+    let barrier = match msg {
+        DispatcherMessage::Chunk(c) => {
+            return Ok(Message::Chunk(c));
+        }
+        DispatcherMessage::Barrier(b) => b,
+        DispatcherMessage::Watermark(watermark) => {
+            return Ok(Message::Watermark(watermark));
+        }
+    };
+    let mutation_subscriber = get_mutation_subscriber(&barrier);
+
+    let mutation = mutation_subscriber
+        .recv()
+        .await
+        .ok_or_else(|| anyhow!("failed to receive mutation of barrier {:?}", barrier))
+        .map(|(prev_epoch, mutation)| {
+            assert_eq!(prev_epoch, barrier.epoch.prev);
+            mutation
+        })?;
+    Ok(Message::Barrier(Barrier {
+        epoch: barrier.epoch,
+        mutation,
+        kind: barrier.kind,
+        tracing_context: barrier.tracing_context,
+        passed_actors: barrier.passed_actors,
+    }))
+}
+
 impl LocalInput {
-    pub fn new(channel: Receiver, actor_id: ActorId) -> Self {
+    pub fn new(
+        channel: Receiver,
+        upstream_actor_id: ActorId,
+        self_actor_id: ActorId,
+        local_barrier_manager: LocalBarrierManager,
+    ) -> Self {
         Self {
-            inner: Self::run(channel, actor_id),
-            actor_id,
+            inner: Self::run(
+                channel,
+                upstream_actor_id,
+                self_actor_id,
+                local_barrier_manager,
+            ),
+            actor_id: upstream_actor_id,
         }
     }
 
     #[try_stream(ok = Message, error = StreamExecutorError)]
-    async fn run(mut channel: Receiver, actor_id: ActorId) {
-        let span: await_tree::Span = format!("LocalInput (actor {actor_id})").into();
+    async fn run(
+        mut channel: Receiver,
+        upstream_actor_id: ActorId,
+        self_actor_id: ActorId,
+        local_barrier_manager: LocalBarrierManager,
+    ) {
+        let span: await_tree::Span = format!("LocalInput (actor {upstream_actor_id})").into();
+        let mut mutation_subscriber = None;
         while let Some(msg) = channel.recv().verbose_instrument_await(span.clone()).await {
-            yield msg;
+            yield process_msg(msg, |barrier| {
+                mutation_subscriber.get_or_insert_with(|| {
+                    local_barrier_manager.subscribe_barrier_mutation(self_actor_id, barrier)
+                })
+            })
+            .await?;
```

Comment:
There used to be no need for local exchanges to acquire mutation from the channel because we don't prune it on the sender side. So this seems to be a regression. 🤔

Reply:
There shouldn't be too much regression when we have https://github.com/risingwavelabs/risingwave/pull/17613. When local input receives a barrier, in most cases the mutation already be sent by the `LocalBarrierWorker` to the mutation subscribe channel.